r/R/trackplots.R

```diff
@@ -113,6 +113,92 @@ get_trackplot_height <- function(plot) {
   }
 }
 
+#' Calculate y positions for transcript segments to avoid overlap
+#' Steps:
+#' 1. Calculate the maximum overlap depth of transcripts
+#' 2. Iterate through transcript start/end in sorted order
+#' 3. Randomly assign each transcript a y-coordinate between 1 and max overlap depth,
+#'   with the restriction that a transcript can't have the same y-coordinate
+#' If is a transcript, filter out the exons in determining overlap so there aren't repeated steps
+trackplot_calculate_segment_height <- function(data, is_transcript = FALSE) {
+  data <- tibble::rownames_to_column(data, "idx_col")
+  boundaries <- data %>%
+    dplyr::filter(if(is_transcript) {!!as.symbol("feature") == "transcript"} else {TRUE}) %>%
+    {
+      dplyr::bind_rows(
+        dplyr::transmute(., pos = start, idx_col, start = TRUE),
+        dplyr::transmute(., pos = end, idx_col, start = FALSE)
+      )
+    } %>%
+    dplyr::arrange(pos)
+  
+  if (nrow(data) > 0) total_positions <- max(cumsum(2 * boundaries$start - 1))
+  else total_positions <- 0L
+  y_pos <- integer(0) # Names: tx_id, value: y position for transcript
+  occupied_y <- rep(FALSE, total_positions) # Boolean vector, marking which y positions are open
+  prev_seed <- get_seed()
+  set.seed(12057235)
+  free_positions <- total_positions
+  for (i in seq_len(nrow(boundaries))) {
+    row <- boundaries[i, ]
+    if (row$start) {
+      # if else block to account for sample interpreting int of length 1 as a range
+      if (free_positions == 1) {
+        assigned_pos <- which(!occupied_y)[1]
+      } else {
+        assigned_pos <- sample(which(!occupied_y), 1)
+      }
+      y_pos[row$idx_col] <- assigned_pos
+      occupied_y[assigned_pos] <- TRUE
+      free_positions <- free_positions - 1
+    } else {
+      assigned_pos <- y_pos[row$idx_col]
+      occupied_y[assigned_pos] <- FALSE
+      free_positions <- free_positions + 1
+    }
+  }
+  restore_seed(prev_seed)
+  data <- dplyr::mutate(data, y=y_pos[idx_col])
+  # Give exons the same y position as the transcript with same transcript id
+  if (is_transcript) {
+    data <- data %>% dplyr::group_by(transcript_id) %>%
+      dplyr::mutate(y = replace(y, is.na(y), max(y,na.rm=TRUE))) %>%
+      dplyr::ungroup()
+  }
+  return(data)
+}
+
+#' Break up segments into smaller segments the length of the plot, divided by size
+#' @param data Dataframe of full segments to be broken up
+#' @param region Region to be plotted with end and start attr
+#' @param size int Number of arrows to span the x axis of track
+#' @param extra_cols Vector of characters of columns to keep from data
+trackplot_create_arrow_segs <- function(data, region, size = 150, extra_cols = c()) {
```

Comment:
Agreed, I was thinking that it could be more elegant to subset and not create duplicated metadata information if not needed.  Pragmatically the space used shouldn't be that much either.

Reply:
My understanding of R is that if you don't modify the values in a data column, the storage space will be shared (copy-on-write). So then the only additional space would be for things like column names, not even for the duplicated data. (Though this space savings would potentially break if we were doing subsetting or any other modification of one copy downstream)